### PR TITLE
chore: refactor helpers common code into a pkg

### LINKED ISSUE #122: chore: convert pipeline images from shell scripts to Go binaries
We have a couple of binaries that we run in Pipelines: `reader` and `update-status`. These are currently just shell scripts, but because of this, they're not very testable.

Currently, they're only covered in our system tests, which can't/shouldn't cover every edge case, so some testing is missing (and issues like #118 can occur).

This issue is to rewrite the scripts in Go, and add some unit tests, to ensure test coverage of these scripts.

(Note: There may be more scripts that we could convert too - `reader` and `update-status` are the two which seem most important, but if you find others, either raise a new issue or convert them as part of this issue depending on how big they are!)

### PR DESCRIPTION
**Pull request by kirederik**
Refactors the reader pipeline component into a testable library and adds unit tests

- Extracts reader functionality into a dedicated library package
- Introduces shared helpers for kubernetes client and parameter handling
- Adds unit tests for the reader component

closes #122

### CODE REVIEW

**Comment by kirederik** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/syntasso/kratix/313?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#313** <a href="https://app.graphite.dev/github/pr/syntasso/kratix/313?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/syntasso/kratix/313?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#312** <a href="https://app.graphite.dev/github/pr/syntasso/kratix/312?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: work-creator/pipeline/cmd/update-status/main.go [hunk 1]
```diff
@@ -63,70 +49,21 @@ func run() error {
...
+	if params.IsLastPipeline {
 		mergedStatus = lib.MarkAsCompleted(mergedStatus)
 	}
 
 	// Apply merged status to the existing object
 	existingObj.Object["status"] = mergedStatus
 
 	// Update the object's status
-	if _, err = dynamicClient.UpdateStatus(context.TODO(), existingObj, metav1.UpdateOptions{}); err != nil {
+	if _, err = objectClient.UpdateStatus(context.TODO(), existingObj, metav1.UpdateOptions{}); err != nil {
```
Comment (ChunyiLyu) [inline work-creator/pipeline/cmd/update-status/main.go:60]:
```suggestion
	if _, err = objectClient.UpdateStatus(context.Background(), existingObj, metav1.UpdateOptions{}); err != nil {
```

File: work-creator/pipeline/lib/reader.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		return fmt.Errorf("failed to write object to file: %v", err)
+	}
+
+	return nil
+}
+
+func writePromiseToFile(client dynamic.Interface, params *helpers.Parameters) error {
+	promiseClient := client.Resource(helpers.PromiseGVR())
+
+	obj, err := promiseClient.Get(context.TODO(), params.PromiseName, metav1.GetOptions{})
```
Comment (ChunyiLyu) [inline work-creator/pipeline/lib/reader.go:54]:
```suggestion
	obj, err := promiseClient.Get(context.Background(), params.PromiseName, metav1.GetOptions{})
```

**Review by ChunyiLyu — APPROVED** [review]

**Comment by kirederik** [comment]
### Merge activity

* **Jan 3, 11:00 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/syntasso/kratix/313).
